scala/scala-impl/src/org/jetbrains/plugins/scala/lang/refactoring/move/anonymousToInner/ScalaAnonymousToInnerHandler.scala

```diff
@@ -0,0 +1,162 @@
+    implicit val projectContext: ProjectContext = new ProjectContext(project)
+
+    val args = variables.map(_.variable.getName).mkString(", ")
+    val text = s"""new $name($args)""".stripMargin
+
+    createElementFromText[ScNewTemplateDefinition](text, newTemplate)
+  }
+
+  private def collectUsedVariables(anonClass: ScExtendsBlock): Array[VariableData] = {
+    var res: List[ScParameter] = Nil
+    anonClass.accept(new ScalaRecursiveElementVisitor() {
+      override def visitReferenceExpression(expression: ScReferenceExpression) = {
+        val refElement = expression.resolve
+        refElement match {
+          case p: ScParameter =>
+            val containingClass = PsiTreeUtil.getParentOfType(p, classOf[ScClass])
+            val containingObject = PsiTreeUtil.getParentOfType(p, classOf[ScObject])
+            val isTopLevelScala3Definition = containingObject == null && containingClass == null
+            if (PsiTreeUtil.isAncestor(containingClass, anonClass, true) || PsiTreeUtil.isAncestor(containingObject, anonClass, true) || isTopLevelScala3Definition) {
+              res = p :: res
+            }
+          case _ =>
+        }
+        super.visitReferenceExpression(expression)
+      }
+    })
+    res.reverse.distinct.map(parameter =>
+      new ScalaVariableData(parameter, true, parameter.`type`().getOrNothing)
+    ).toArray
+
+  }
+
+  private case class DialogResult(className: String, variables: Array[VariableData])
+
+  private def showRefactoringDialog(project: Project, anonClass: ScExtendsBlock, usedVariables: Array[VariableData], target: Either[ScFile, ScTemplateDefinition]) = {
+    val dialog = new ScalaAnonymousToInnerDialog(project, anonClass, usedVariables, target)
+    if (dialog.showAndGet()) {
+      Some(DialogResult(dialog.getClassName, dialog.getVariables))
+    } else
+      None
+  }
+
+  private def createClass(name: String, anonClass: ScExtendsBlock, variables: Array[VariableData], project: Project): ScClass = {
+    implicit val projectContext: ProjectContext = new ProjectContext(project)
+
+    val parametersText = variables.map(v => s"${v.name}: ${v.`type`.getPresentableText}").mkString(", ")
+    val parameterOldNameToNewName = variables.map(v => v.variable.getName -> v.name).toMap
+
+    val text = s"class $name($parametersText) extends ${anonClass.getText}"
+
+    val newClass = createElementFromText[ScClass](text, anonClass)
+
+    newClass.accept(new ScalaRecursiveElementVisitor() {
+      override def visitReferenceExpression(expression: ScReferenceExpression) = {
+        val refElement = expression.resolve
+        refElement match {
+          case p: ScParameter =>
+            expression.handleElementRename(parameterOldNameToNewName(p.getName))
+          case _ =>
+        }
+        super.visitReferenceExpression(expression)
+      }
+    })
+
+    newClass
+  }
+
+  @tailrec
+  private def findAnonymousClass(element: PsiElement): Option[ScExtendsBlock] =
```

Comment:
thanks for noticing this, this entire function is unnecessary since I now pass it ScNewTemplateDefinition anyway (it wasn't like that before). so I'll delete it